src/timecode.rs

```diff
@@ -497,17 +498,122 @@ impl Timecode {
     ```rust
     # use vtc::{Timecode, rates};
     let tc = Timecode::with_frames("01:00:00:00", rates::F23_98).unwrap();
-    assert_eq!("5400+00", tc.feet_and_frames())
+    assert_eq!("5400+00", tc.feet_and_frames_35mm_4p())
     ```
     */
-    pub fn feet_and_frames(&self) -> String {
-        let result = div_mod_floor(abs(self.frames()), FRAMES_PER_FOOT);
-        let feet = result.0;
-        let frames = result.1;
+    pub fn feet_and_frames_35mm_4p(&self) -> String {
+        self.feet_and_frames_impl(4, PERFS_PER_FOOT_35)
+    }
+
+    /**
+     Returns the feet and frames representation in 16mm footage.
+
+     # What it is
+
+     On 16mm film, there are forty frames of film in each foot, one perforation
+     per frame. However, 16mm film is edge coded every six inches, with twenty
+     frames per code, so the footage "1+19" is succeeded by "2+0".
+
+     # Where you see it
+
+     16mm telecines, 16mm edge codes.
+
+     # Examples
+
+    ```rust
+    # use vtc::{Timecode, rates};
+    let tc = Timecode::with_frames("00:00:01:00", rates::F23_98).unwrap();
+    assert_eq!("1+04", tc.feet_and_frames_16mm())
+    ```
+     */
+    pub fn feet_and_frames_16mm(&self) -> String {
+        self.feet_and_frames_impl(1, PERFS_PER_6INCHES_16)
+    }
+
+    /**
+    Returns the feet and frames representation in 35mm, 2-perf footage.
+
+    # What it is
+
+    35mm 2-perf film records 32 frames in a foot of film, instead of the usual 16.
+    This creates a negative image with a wide aspect ratio using standard spherical
+    lenses and consumes half the footage per minute running time as standard 35mm,
+    while having a grain profile somewhat better than 16mm while not as good as
+    standard 35mm.
+
+    # Where you see it
+
+    35mm 2-perf formats are uncommon though still find occasional use, the process is
+    usually marketed as "Techniscope", the original trademark for Technicolor Italia's
+    2-perf format. It was historically very common in the Italian film industry prior
+    to digital filmmaking, and is used on some contemporary films to obtain a film look
+    while keeping stock and processing costs down.
+
+    # Example
+    ```rust
+    # use vtc::{Timecode, rates};
+    let tc = Timecode::with_frames("00:00:01:00", rates::F23_98).unwrap();
+    assert_eq!("0+24", tc.feet_and_frames_35mm_2p());
+    let tc2 = Timecode::with_frames("00:00:02:00", rates::F23_98).unwrap();
+    assert_eq!("1+16", tc2.feet_and_frames_35mm_2p());
+    ```
+    */
+    pub fn feet_and_frames_35mm_2p(&self) -> String {
```

Comment:
And a parser for this one as well.

Reply:
How do you want to integrate the footage parsing into the `&str::to_frames()` logic? Footage representations can be ambiguous and the Framerate struct doesn't give enough context to pick.

Reply:
Hmmm. I see what you're saying.

I think the best option would be to add a `::to_frames(rate, opts)` function to the trait that takes a new `Frames::Opts` struct. That options struct can have a `feet_frames_format` field that can hold the format we're parsing using the same enum as the renderer.

Then we move the existing logic from `::to_frames(rate)`  into `::to_frames(rate, opts)`, and instead just pass a default opts object into it with the 35mm, 4-perf format used as the default interpretation.

Let me know if you want me to make my own PR setting up that machinery that you can rebase on top of. That's a lot of internal guts work for this feature.

Reply:
Ignore that. Better idea:

Declare new types for feet and frames that the caller can cast their string to first. Then implement the parsers for the new formats on those new types.

So make this type:

```rust
pub struct FeetFramesStr {
    input: &str,
    format: FeetFrames
}
```

And implement a constructor and the `Frames` trait for it.

Then you could do the following:

```rust
input = FeetFramesStr("00+06", FeetFrames::FF16mm):

tc = Timecode::with_frames(input);
```

That fits much more cleanly with the ethos of the design of this lib: you can add support for any format you want via traits. Then we keep the existing &str implementation that assumes 35mm, 4-perf by default. 

Apologies if any of my code is off. I'm on mobile.

Reply:
Looks fine, the constructor of FeetFramesStr should return a result type in case the footage string doesn't parse (and it's going to capture a lifetime because of that &str ick) but yes this can be done.

What if we did a builder-type deal so you could do something like...  

```rust

let one_minute = Timecode::from_frames("90+0".feet_frames(FeetFramesRep::FF35mm4perf)?, rate)
```

We'd have to add a method to the impl of &str to create either a mezzanine struct or just crunch it to a frame count then and there. At least this way we're not introducing a new struct and holding onto a borrowed string.

Reply:
I don't think the constructor should do any parsing itself, it should purely be a place to park the data for `Timecode::with_frames` to parse it through the `Frames` trait, which should be the thing that does the actual parsing. That impl will be responsible for returning the result. The struct should just be a short-lived adapter for that's dropped as soon as `Timecode::with_frames` does it's thing. I don't see a ton of issues with holding onto a borrow inside the struct. If someone really can't spare it they can clone the source string first no?

Make sense?

Reply:
Also just realized I meant `FramesSource` trait instead of `Frames`. It's been a second since I've been in this lib haha.

Reply:
Sorry I didn't see the sentence where you were proposing a builder. That's what I get for looking while being tired. I'll remove my earlier comment.

I don't think I'm inclined to go that route. I don't mind holding onto read borrow for the parsing step, the caller can always clone first if they can't spare it. If lifetimes become too cumbersome we could always make the field a non-borrowed `String`, but I would like to see if we can avoid a forced copy like that.

Builder pattern just doesn't line up with how the lib is currently written, and I'd like to avoid doing something bespoke for F+F unless there is no other choice. API consistency means a friendlier interface for new users. Right now that means the `FramesSource` impl should be the thing doing the parsing.

I think it makes the user-facing API a little cleaner as well, you don't have to handle a return value from the type adapter, just the actual `::with_frames` function.

I'm more than happy to stub out the logic flow for you and grapple with the lifetime stuff. Then you can fill in the actual parsing logic, just let me know!

Reply:
In 2914c8c, review the examples in lib.rs to see if this works okay, I did a new struct but I did implement an `Into` trait and extension on str to make constructing them more brief.

We have to rewrite the unit test machinery too though.

Reply:
Thanks for this! Looks good overall, I have a couple notes but I think we are zeroing in on the final approval:

Let's call `FeetFramesFormat` `FilmFormat` instead. More compact and better captures what it is without tying it to F+F specifically.

I'd like `FeetFrames` to be called `FeetFramesStr` — I think that makes it a little more clear that it's just an adapter that holds a string + some data, rather than being some esoteric type.

Lastly, I'd rather have a `new` constructor instead of relying on the `Into` trait. `Into` seems the most useful when using it to auto-cast on function calls and the like. Here, I think it adds complexity without much upside, as nowhere in the library's code will we be able to use it to auto-cast, especially since it needs to take an arg.

In my opinion, this:

```rust
Timecode::with_frames(FeetFrameStr::new("00+04", FilmFormat::FF16mm))?
```

Is more explicit than and easier to reason about than:

```rust
Timecode::with_frames("00+04".to_feet_frames(FilmFormat::FF16mm))?
```

The first is only four characters more than using `Into`, and you don't get to see the actual type involved at-a-glance.

But I'm also fairly new to Rust, so I don't always have an intuitive sense for all the lifetime and borrow implications of a piece of code; if you feel like using a constructor will be limiting in some way, let me know and I'll reconsider.

Reply:
Okay I think that's everything, easy peasy. Also saw some commits in the upstream that I've merged in.

ec95bb6